tests/test_app.py

```diff
@@ -416,27 +430,97 @@ def test_only_running_once():
     """
     If we are the first to acquire the application lock we should succeed
     """
+    setup_exception_handler()
     check_only_running_once()
+    _shared_memory.release()
     assert True
 
 
 def test_running_twice():
+    # try chaining instead of timing based stuff
+
+    # Comment on the original test says:
+    #
+    #   It's important that the two competing processes are not part of the same
+    #   process tree; otherwise the second attempt to acquire the mutex will
+    #   succeed (which we don't want to happen for our purposes)
+    #
+    # However, on macOS it always seems to have the same parent process id and group
+    # process id and this test seems to work on all platforms on CI.
+    #
+    cmd2 = "".join(
+        (
+            "-c",
+            "from mu import app;",
+            "app.setup_exception_handler();",
+            "app.check_only_running_once()"
+            # should throw an exception and exit with code 2 if it's already running
+        )
+    )
+
+    cmd1 = "".join(
+        (
+            "-c",
+            "import subprocess;",
+            "import sys;",
+            "from mu import app;",
+            "app.setup_exception_handler();",
+            "app.check_only_running_once();",
+            # launch child process that will try to 'launch' the app again:
+            'child2 = subprocess.run([sys.executable, "{0}"]);'.format(cmd2),
+            # clean up and exit returning child process result code (which should be 2
+            # if this 'already running' code is working)
+            "app._shared_memory.release();",
+            "exit(child2.returncode)",
+        )
+    )
+
+    child1 = subprocess.run([sys.executable, cmd1])
+    print("child 1 return code: {}".format(child1.returncode))
+    assert child1.returncode == 2
+
+
+def test_running_twice_after_exception():
+    """
+    If we run, throw an exception and then run again we should succeed.
+    """
+    # call test that causes app to throw an exception by running it twice
+    test_running_twice()
+    # test that we can still run after exception thrown
+    test_only_running_once()
+
+
+def test_running_twice_after_generic_exception():
```

Comment:
New test that tests the exception handler clears the app shared memory semaphore so that the app can be launched again after an exception.